ludwig/schema/config_object.py

```diff
@@ -0,0 +1,82 @@
+from ludwig.constants import (
+    COMBINER,
+    DECODER,
+    ENCODER,
+    HYPEROPT,
+    INPUT_FEATURES,
+    NAME,
+    OUTPUT_FEATURES,
+    PREPROCESSING,
+    TRAINER,
+    TYPE,
+)
+from ludwig.schema.combiners.base import BaseCombinerConfig
+from ludwig.schema.combiners.concat import ConcatCombinerConfig
+from ludwig.schema.features.utils import input_type_registry, output_type_registry
+from ludwig.schema.preprocessing import PreprocessingConfig
+from ludwig.schema.trainer import BaseTrainerConfig, ECDTrainerConfig
+
+
+class InputFeatures:
+    """InputFeatures is a container for all input features."""
+
+    pass
+
+
+class OutputFeatures:
+    """OutputFeatures is a container for all output features."""
+
+    pass
+
+
+class Config:
+    """This class is the implementation of the config object that replaces the need for a config dictionary
+    throughout the project."""
+
+    input_features = InputFeatures()
+    output_features = OutputFeatures()
+    combiner: BaseCombinerConfig = ConcatCombinerConfig()
+    trainer: BaseTrainerConfig = ECDTrainerConfig()
+    preprocessing = PreprocessingConfig()
+    hyperopt = {}
+    defaults = {}
+
+    def __init__(self, config_dict):
+        self.parse_input_features(config_dict[INPUT_FEATURES])
+        self.parse_output_features(config_dict[OUTPUT_FEATURES])
+
+        if COMBINER in config_dict:
```

Comment:
Hey @connor-mccorm , is this IF here to deal with Trees where we don't have a combiner?

Reply:
Not quite, basically how this Config class will work is it will take all attributes defined in the user defined `config_dict` and update the defaults with them. Since the combiner is an optional section, it may or may not be present in the user config so we have to check to see if it's there to tell if we need to run any updates. This does bring up a good point though since I am only accounting for ECD model trainer here and I need to have GBM support as well. Either way, this file is actually not being used in this PR, it is actually part of the config object PR so I can just remove it from here since it isn't relevant.